### DISCUSSION TITLE
Heartbeat mechanism with Smallrye AMQP connector

### DISCUSSION THREAD

**Discussion by mnguetsa** [quarkusio/quarkus#39365] [Q&A] [answered]
### Describe the bug

Hi everyone,

this is mostly a question as I couldn't change the label. 
Asked the same [here](https://github.com/quarkusio/quarkus/discussions/39246) but did not receive answers so far. 

I am using smallrye reactive messaging with Quarkus to connect to an amqp message broker.
However after a certain idle timeout my application loses connection to the broker due to inactivity and has to reconnect.
I would like to keep the connection without reconnection.
I tried to customize the underlying amqp client by setting TCP KeepAlive to true but it does not help with connection losses.
https://quarkus.io/guides/amqp-reference#customizing-the-underlying-amqp-client

My assumption here is that the idle timeout is set at the broker side because I don't think the smallrye amqp has an idle timeout right?

Ideally there should be a mechanism for the client to keep the connection like a heartbeat mechanism in case the broker has a timeout. I could not find something like that in the doc for Smallrye AMQP.
Therefore I need your help in case you know how to solve this.
Thanks

### Output of `java -version`

21

### Quarkus version or git rev

3.8.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

gradle 8.5

[a comment by an automated account is not shown]

**Comment by mnguetsa** [comment]
@geoand @cescoffier Any idea how I could do this? 😬

**Comment by geoand** [comment]
cc @ozangunalp

**Comment by cescoffier** [comment]
No, except modifying the AMQP client options, I do not see another approach. 

I would disable the health check and implement a custom one instead.

**Comment by mnguetsa** [comment]
> No, except modifying the AMQP client options, I do not see another approach.
> 
> I would disable the health check and implement a custom one instead.

Hi @cescoffier 
I tried modifying the clients options too. But I realized setTcpKeepAlive(true) had no effect on keeping the connection. Or is there another option I missed in the AMQP client options? So far this is what I did with the AMQP client Options:

```java
@ApplicationScoped
public class AmqpConfigurationProvider {
  @ConfigProperty(name = "amqp-port")
  int port;

  @ConfigProperty(name = "amqp-use-ssl")
  boolean useSsl;

  @Produces
  @Identifier("custom-amqp-client-options")
  public AmqpClientOptions getNamedOptions() {
    return new AmqpClientOptions().setTcpKeepAlive(true).setPort(port).setSsl(useSsl);
  }
}
```

I would use the health check deactivation as last solution. If you know another AMQP option I can try please let me know. 
Thanks a lot

**Comment by cescoffier** [comment]
Which broker are you using? Do they use an idle timeout? Could you potentially send fake frames regularly to keep it open?

**Reply by mnguetsa** [reply]
I am connecting to Solace PubSub through AMQP 1.0. Also the timeout is very regular, always at the same interval of time  as per my app logs therefore I have the strong assumption that an idle timeout is set at the broker side. 

How to send those fake frames with Smallrye reactive Messaging within Quarkus? I haven't seen that in the docs yet.

**Reply by ozangunalp** [reply]
I read here that Solace implementation honours the keep-alive: https://docs.solace.com/API/AMQP/AMQP-Protocol-Conformance.htm#Sec2-4-5

**Reply by ozangunalp** [reply]
FYI there is a Quarkus extension being developed by the Solace team for Solace PubSub protocol (not AMQP 1.0) : https://github.com/SolaceLabs/solace-quarkus/

**Reply by mnguetsa** [reply]
Yes I read this as well but it sounded more like a guideline than what really happens under the hood. Then I tried the tcp keep alive and saw no difference. 
However now I am gravitating towards implementing my own heartbeat mechanism and was wondering whether you know how I could send some empty/fake frames with Quarkus and Smallrye reactive messaging?

**Reply by mnguetsa** [reply]
> FYI there is a Quarkus extension being developed by the Solace team for Solace PubSub protocol (not AMQP 1.0) : https://github.com/SolaceLabs/solace-quarkus/

Interesting. Thanks a lot for the link. But for now I can only use AMQP 1.0

**Reply by ozangunalp** [reply]
Did you try also setting the idle timeout on AMQP client options ? As per the Solace doc they should honor it and not close the connection. 
Otherwise, if I am not mistaken, reactive messaging AMQP connector doesn't expose the current amqp connection.

**Reply by mnguetsa** [reply]
> Did you try also setting the idle timeout on AMQP client options ? 

I looked at the default in Vert.x AmqpClientOptions: https://vertx.io/docs/apidocs/io/vertx/core/net/TCPSSLOptions.html#DEFAULT_IDLE_TIMEOUT
It seems the default is already 0. I thought this would mean there is no timeout. Or am I mistaking and should set something?

**Reply by ozangunalp** [reply, marked as answer]
I am not sure how the Vert.x Proton interprets the options. Maybe it is not even the idle timeout but the `heartbeat` property: 
https://github.com/vert-x3/vertx-proton/blob/dd39eeea85764255750fe19013ad231f8dee4c16/src/main/java/io/vertx/proton/impl/ProtonTransport.java#L75

The proton client will then advertise the idle timeout opening a transport to the remote peer.

**Reply by mnguetsa** [reply]
Hmm From the ProtonClientOptions#setHeartbeat(...) Javadoc I am not sure whether it is what we need here
https://github.com/vert-x3/vertx-proton/blob/dd39eeea85764255750fe19013ad231f8dee4c16/src/main/java/io/vertx/proton/ProtonClientOptions.java#L425

**Reply by ozangunalp** [reply]
That's why I pointed you to the TCP idle timeout config but looking at the proton transport code, it does set the AMQP idle timeout from the heartbeat config.

**Reply by mnguetsa** [reply]
Ok thanks I will try that out and let you know. That would mean I will have to set it pretty high. Because I actually need a permanent connection

**Reply by mnguetsa** [reply]
Unfortunately setting the heartbeat on the AmqpClientOptions did not help either 

```java
return new AmqpClientOptions()
        .setTcpKeepAlive(true)
        .setPort(port)
        .setSsl(useSsl)
        .setHeartbeat((int) TimeUnit.DAYS.toMillis(30));
```

**Comment by ozangunalp** [comment]
I think AMQP 1.0 broker in Azure (I don't remember the product name) had a similar behaviour.

**Comment by ozangunalp** [comment]
Probably, you are sending the TCP keep-alive but the server doesn't respect it and closes the connection anyway.

**Reply by mnguetsa** [reply]
Yes

**Comment by cescoffier** [comment]
Cloud-based broker tends to cut connections quickly if there are no L7 frames.

[a comment by an automated account is not shown]

**Comment by mnguetsa** [comment]
@ozangunalp I have some good news. Setting the heartbeat finally helped. But I have to explain. Here is what the javadoc of the setHeartBeat method on the ProtonClientOptions says: **_Set the heartbeat (in milliseconds) as maximum delay between sending frames for the remote peers. If no frames are received within 2*heartbeat, the connection is closed_**
Reading that doc I first thought I need to set that value high so it does not disconnect and it did not help with reconnection. But in contrary when I then tried with a very low value I couldn't see disconnection anymore. It works even without setting TCP keep Alive. What really happens under the hood I don't know. If you have an explanation I am open to learn.

In summary maybe it is something we can add to the Quarkus doc in the section about  customizing the underlying AMQP client. I could also submit a PR if you like. If you would like to do it by yourself it is also fine. 
I really appreciate your help thanks a lot. 

```java
return new AmqpClientOptions()
        .setTcpKeepAlive(true)
        .setPort(port)
        .setSsl(useSsl)
        .setHeartbeat(3000);
```

**Reply by ozangunalp** [reply]
That's very good news! Please submit a PR explaining that, it can serve others

**Comment by mnguetsa** [comment]
Here is the PR I just created for that in Quarkus doc : https://github.com/quarkusio/quarkus/pull/39406

**Reply by mnguetsa** [reply]
PR for adjusting similar doc in smallrye-reactive-messaging: https://github.com/smallrye/smallrye-reactive-messaging/pull/2528

**Comment by mnguetsa** [comment]
If you know someone at Qaurkus repo that could review the PR I would really appreciate: https://github.com/quarkusio/quarkus/pull/39406
Thank you

**Reply by ozangunalp** [reply]
Sorry I missed that, it is merged now.
